unblob/handlers/archive/zip.py

```diff
@@ -98,78 +98,76 @@
-    # chunk within our firmware image, and then carve that chunk out.
-    # Then, we make this is a BytesIO stream, so we can just pass this
-    # stream to ZipFile.
-
-    file_names = set()
-    encrypted_files = set()
-    zip_end = _find_end_of_zip(file, start_offset)
+def calculate_zipfile_end(file: io.BufferedIOBase, start_offset: int) -> int:
+    # If we just pass a firmware blob with multiple ZIP files in it to zipfile.ZipFile, it seems
+    # that it will basically scan for the final EOCD record header, and assume that that's where
+    # the file ends.
+    # E.g. in the case our firmware image looks like this:
+    # | ZIPFILE | SOMETHING ELSE | ZIPFILE |
+    # zipfile.ZipFile() will assume:
+    # |    THIS IS ALL THE SAME ZIPFILE    |
+    # For obvious reasons, this is not helpful in our case. We need to try to guess the length of
+    # the ZIP file chunk within our firmware image, independently, and then carve that chunk out.
 
     file.seek(start_offset)
-    content = io.BytesIO(file.read(zip_end - start_offset))
 
-    with ZipFile(content) as z:
-        logger.info("Found ZIP filenames", filenames=[x.filename for x in z.infolist()])
-        for g in z.infolist():
-            if g.flag_bits & 0b0001:
-                encrypted_files.add(g.filename)
-            file_names.add(g.filename)
+    # In our case, we want to find the first instance of the EOCD record header, not the last!
+    zip_end = find_first(file, b"\x50\x4b\x05\x06") + start_offset
 
-    size = zip_end - start_offset
-    return size
+    if zip_end == -1:
+        return -1
+    else:
+        file.seek(zip_end)
+        _ = cparser.end_of_central_directory(file)
+        return file.tell()
 
 
 def calculate_chunk(
     file: io.BufferedIOBase, start_offset: int
 ) -> Union[ValidChunk, UnknownChunk]:
+
     header = cparser.local_file_header(file)
     if header.version_needed_to_extract > MAXIMUM_VERSION:
         return UnknownChunk(
             start_offset=start_offset,
             reason=f"ZIP (0x{start_offset:x}): Version too high!",
         )
 
-    size = _guess_zip_size(file, start_offset)
+    end_of_zip = calculate_zipfile_end(file, start_offset)
+
+    if end_of_zip == -1:
+        return UnknownChunk(
+            start_offset=start_offset,
+            reason="Couldn't find the EOCD magic bytes.",
+        )
+
+    file.seek(start_offset)
+
+    encrypted_files = set()
+    all_files = set()
+
+    this_zip_chunk = io.BytesIO(file.read(end_of_zip - start_offset))
+    with zipfile.ZipFile(this_zip_chunk) as zip:
+        for zipinfo in zip.infolist():
+            if zipinfo.flag_bits & 0b0001:
+                encrypted_files.add(zipinfo.filename)
+            all_files.add(zipinfo.filename)
+
+    if len(encrypted_files) > 0:
+        # TODO: We can't handle encrypted ZIP files yet, so we fall back to the UnknownChunk in the
+        # cases where there are encrypted files in the ZIP.
+        return UnknownChunk(
```

Comment:
Let's talk about encrypted files within ZIP archives here. With the current logic, an archive with multiple plaintext files and one encrypted file will not be processed by unblob. This is problematic.

If we provide the `-P` argument with an empty value, unzip will simply skip encrypted files with wrong password without triggering exceptions:

```
unzip -d lol -P '' fruits.zip
Archive:  fruits.zip
   skipping: fruits/banana.txt       incorrect password
```

I'd be in favor of using this in order to emit a valid chunk even if an encrypted file within an archive is found. This would make unblob carve out the whole zip so that an analyst can take a deeper look at it.

Given the extraction process won't be lossless in these cases (encrypted files won't be extracted), I'd be in favor of logging a warning message saying "X files from zip chunk Y are encrypted but we don't have a password, better have a look !" or something.